core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,466 @@
+                    GenerateV1StatisticsJob.EVENT_TYPE_STATE_ANSWERS):
+                state_answer_event_dicts.append(value)
+
+        # This dict will store the cumulative values of state stats from version
+        # 1 till the latest version of an exploration.
+        state_stats_mapping = {}
+        # Sort list of state hit events in increasing order of timestamp.
+        state_hit_event_dicts = sorted(state_hit_event_dicts, key=lambda e: e[
+            'created_on'])
+        for version in range(1, latest_exp_version+1):
+            start_exploration_events_for_this_version = [
+                val for val in start_exploration_event_dicts if val[
+                    'version'] == version]
+            complete_exploration_events_for_this_version = [
+                val for val in complete_exploration_event_dicts if val[
+                    'version'] == version]
+            state_hit_events_for_this_version = [
+                val for val in state_hit_event_dicts if val[
+                    'version'] == version]
+            state_answer_events_for_this_version = [
+                val for val in state_answer_event_dicts if val[
+                    'version'] == version]
+            versioned_exploration = exp_services.get_exploration_by_id(
+                exp_id, version=version)
+
+            is_revert = False
+            if version == 1:
+                # Create default state stats mapping for the first version.
+                for state_name in versioned_exploration.states:
+                    state_stats_mapping[state_name] = (
+                        stats_domain.StateStats.create_default())
+            else:
+                exp_commit_log = exp_models.ExplorationCommitLogEntryModel.get(
+                    'exploration-%s-%s' % (exp_id, version))
+                change_list = exp_commit_log.commit_cmds
+
+                # Handling state additions, renames and deletions.
+                for change_dict in change_list:
+                    if change_dict['cmd'] == exp_domain.CMD_ADD_STATE:
+                        state_stats_mapping[change_dict['state_name']] = (
+                            stats_domain.StateStats.create_default())
+                    elif change_dict['cmd'] == exp_domain.CMD_DELETE_STATE:
+                        state_stats_mapping.pop(change_dict['state_name'])
+                    elif change_dict['cmd'] == exp_domain.CMD_RENAME_STATE:
+                        state_stats_mapping[change_dict['new_state_name']] = (
+                            state_stats_mapping.pop(
+                                change_dict['old_state_name']))
+                    elif change_dict['cmd'] == 'AUTO_revert_version_number':
+                        revert_to_version = change_dict['version_number']
+                        is_revert = True
+
+            # Computing num_starts and num_completions.
+            num_starts += len(start_exploration_events_for_this_version)
```

Comment:
If this is the only thing you're going to take away from the event dict, then maybe just keep track of only the counter at the outset? No need to make a list of all "start event" dicts, etc.

Similar comments apply for other fields. If you can save space, try to do so. And the fewer times you iterate through the entire dataset, the better, so try and pick a data structure that allows you to easily update as much stuff as possible on-the-fly in a single pass through all the data at the outset.